### PR TITLE
WIP: Support benchmark in run_e2e.go and jenkins

### PR DESCRIPTION
**Pull request by coufon**
This PR contains the following changes:
1. Move the flag 'ginkgoFlags' from e2e_remote.go to run_e2e.go and pass it as an argument to e2e_remote.go.
1. Modify run_e2e.go to support benchmark mode. In benchmark mode it creates several instances of different GCE machine types/images and run tests in parallel.
2. Add benchmark configuration 'e2e_node/jenkins/benchmark/benchmark-config.yaml'. It specifies a set of tests for benchmark, each test specifies machine/image and ginkgo focus. Host can be specified to run the test on a target instance, it is not implemented in run_e2e.go yet.
3. Add e2e-node-benchmark.properties.
4. Add script e2e-node-benchmark-jenkins.sh.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30425)

### CODE REVIEW

**Comment by coufon** [comment]
@Random-Liu  @dchen1107

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties [hunk 1]
```diff
@@ -0,0 +1,9 @@
+GCE_HOSTS=
+GCE_IMAGE_CONFIG_PATH=test/e2e_node/jenkins/benchmark/benchmark-config.yaml
+GCE_ZONE=us-central1-f
+GCE_PROJECT=test-kube-node-1348
+CLEANUP=true
+GINKGO_FLAGS='--skip="\[Flaky\]"'
+SETUP_NODE=true
+TEST_ARGS=--cgroups-per-qos=true
```
Comment (Random-Liu) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: s/true/false. Currently, we always disable this one.

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties [hunk 2]
```diff
@@ -0,0 +1,9 @@
+GCE_HOSTS=
+GCE_IMAGE_CONFIG_PATH=test/e2e_node/jenkins/benchmark/benchmark-config.yaml
+GCE_ZONE=us-central1-f
+GCE_PROJECT=test-kube-node-1348
```
Comment (Random-Liu) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: ? This should be changed to https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/jenkins/jenkins-ci.properties#L4.

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+CLEANUP=true
+GINKGO_FLAGS='--skip="\[Flaky\]"'
+SETUP_NODE=true
```
Comment (Random-Liu) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: Use `true` only when you are testing it yourself. Change to `false`.

File: test/e2e_node/jenkins/benchmark/benchmark-config.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+---
+images:
+  containervm:
+    image: e2e-node-containervm-v20160321-image
+    project: kubernetes-node-e2e-images
+    machine: n1-standard-1
+    tests:
+      - ".*create 10 pods with 0s? interval"
```
Comment (Random-Liu) [inline test/e2e_node/jenkins/benchmark/benchmark-config.yaml]: wired ending character.

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+TEST_ARGS=--cgroups-per-qos=true
+PARALLELISM=1
```
Comment (Random-Liu) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: ditto.

File: test/e2e_node/runner/run_e2e.go [hunk 6]
```diff
@@ -56,17 +58,24 @@ var deleteInstances = flag.Bool("delete-instances", true, "If true, delete any i
...
 
-var computeService *compute.Service
+const (
+	defaultMachine = "n1-standard-1"
+)
+
+var (
+	computeService *compute.Service
+	arc            Archive
+	globalID       uint64 = 0
```
Comment (Random-Liu) [inline test/e2e_node/runner/run_e2e.go]: Offline discussed, globalID doesn't work.

File: test/e2e_node/runner/run_e2e.go [hunk 7]
```diff
@@ -83,6 +92,9 @@ type TestResult struct {
 //       short-name:
 //         image: gce-image-name
 //         project: gce-image-project
+
```
Comment (Random-Liu) [inline test/e2e_node/runner/run_e2e.go]: Add comments about new files in the image config.

File: test/e2e_node/runner/run_e2e.go [hunk 8]
```diff
@@ -95,6 +107,9 @@ type GCEImage struct {
 	// Defaults to using only the latest image. Acceptible values are [0, # of images that match the regex).
 	// If the number of existing previous images is lesser than what is desired, the test will use that is available.
 	PreviousImages int `json:"previous_images, omitempty"`
+	// The test is regarded as 'benchmark' is Machine is non-empty
```
Comment (Random-Liu) [inline test/e2e_node/runner/run_e2e.go]: Other test could also specify machine type.

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties (hunk 1 above)
Comment (coufon) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: ACK

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties (hunk 2 above)
Comment (coufon) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: ACK

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties (hunk 3 above)
Comment (coufon) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: ACK

File: test/e2e_node/jenkins/benchmark/benchmark-config.yaml (hunk 4 above)
Comment (coufon) [inline test/e2e_node/jenkins/benchmark/benchmark-config.yaml]: ACK

File: test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties (hunk 5 above)
Comment (coufon) [inline test/e2e_node/jenkins/benchmark/jenkins-benchmark.properties]: ACK

File: test/e2e_node/runner/run_e2e.go (hunk 7 above)
Comment (coufon) [inline test/e2e_node/runner/run_e2e.go]: ACK

File: test/e2e_node/runner/run_e2e.go (hunk 8 above)
Comment (coufon) [inline test/e2e_node/runner/run_e2e.go]: ACK. Use non-empty test list as the indicator of benchmark instead.

**Comment by Random-Liu** [comment]
*[test/e2e_node/runner/run_e2e.go, line 571 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30425#-KPPWFRq1KzlFihlOCci:-KPPWFRrWSP3yJ-lL4aP:bothhvt) ([raw file](https://github.com/kubernetes/kubernetes/blob/a9b1594b57c114fc5c64deff96e1ff7506e449f0/test/e2e_node/runner/run_e2e.go#L571)):*
> ```Go
> 	}
> 	atomic.AddUint64(&globalID, 1)
> 	return image.machine + "-" + image.image + "-" + uuid.NewUUID().String()[:8] // strconv.Itoa(int(atomic.LoadUint64(&globalID)))
> ```

Remove globalID and the code in the comment.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30425)*

File: hack/verify-flags/known-flags.txt [hunk 9]
```diff
@@ -508,3 +508,4 @@ windows-line-endings
 www-prefix
 zone-name
 garbage-collector-enabled
+benchmark-config-file
```
Comment (Random-Liu) [inline hack/verify-flags/known-flags.txt]: Not used anymore? :)

File: test/e2e_node/runner/run_e2e.go [hunk 10]
```diff
@@ -534,14 +563,33 @@ func parseInstanceMetadata(str string) map[string]string {
...
 }
 
-func imageToInstanceName(image string) string {
-	return *instanceNamePrefix + "-" + image
+func imageToInstanceName(image *internalGCEImage) string {
+	if image.machine == "" {
+		return *instanceNamePrefix + "-" + image.image
+	}
+	atomic.AddUint64(&globalID, 1)
+	return image.machine + "-" + image.image + "-" + uuid.NewUUID().String()[:8] // strconv.Itoa(int(atomic.LoadUint64(&globalID)))
```
Comment (Random-Liu) [inline test/e2e_node/runner/run_e2e.go]: Remove the extra commetn.

**Comment by Random-Liu** [comment]
Close this one in favor of https://github.com/kubernetes/kubernetes/pull/31039.

**Pull request closed without merging**